modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition_manager.py

```diff
@@ -137,6 +137,63 @@ def get_indices(cls, axis, partitions, index_func=None):
         new_idx = ray.get(new_idx)
         return new_idx[0].append(new_idx[1:]) if len(new_idx) else new_idx
 
+    @classmethod
+    def concat(cls, axis, left_parts, right_parts):
+        result = super(PandasOnRayDataframePartitionManager, cls).concat(
+            axis, left_parts, right_parts
+        )
+        return cls.rebalance_partitions(result)
+
+    @classmethod
+    def rebalance_partitions(cls, partitions):
+        from modin.config import NPartitions
+
+        heuristic = 1.5  # partitions can be 1.5x larger than ideal. Can be modified.
+        if partitions.shape[0] > NPartitions.get() * heuristic:
+            lengths = [[obj._length_cache for obj in row] for row in partitions]
+            # Naive rebalance
+            if any(l is None for row in lengths for l in row):
+                if partitions.shape[0] % NPartitions.get() == 0:
+                    step = partitions.shape[0] // NPartitions.get()
+                else:
+                    step = partitions.shape[0] // NPartitions.get() + 1
+                return np.array(
+                    [
+                        cls.column_partitions(partitions[i : i + step], full_axis=False)
```

Comment:
We probably want to end up implementing actually a composite pattern for `AxisPartition`, meaning that `AxisPartition` has to be a subtype of `BaseDataframePartition` so there won't be a difference for a dataframe of whether to handle a regular or an axis partition.

All of the `BaseDataframePartition` API seems to be trivially implementable for a composite `AxisPartition`.

Reply:
@devin-petersohn In other words, there are fundamentally two ways to go. The first is to do the repartition immediately after the operations that create a suboptimal partition. In this case, it is also necessary to think over the mechanism of deferred repartitioning so as not to perform a cumbersome operation when it can be avoided. The second is to do repartitioning before operations, when we know exactly what we need. In this case, the delayed repartition mechanism is not needed. What do you think?

Reply:
@dchigarev I think it makes sense from a code structure point of view, maybe I am not sure that a virtual partition should inherit from a physical partition.

Reply:
@anmyachev I think it is never good to do the repartitioning physically right away.

Reply:
@devin-petersohn I wonder why the axis_partition class was chosen as the base class for the virtual partition, and not the partition class? For example, the `apply` operation in the case of an axis_partition class must be combined before execution, and then again divided into several parts, which is not always necessary for a virtual partition that should be obtained after rebalancing.
On the other hand, with the current solution, axis_partitions and simple partitions should be interchangeable, which, it seems to me, is not what we want. We want the virtual partition obtained after re-partitioning to behave like a simple partition.

Reply:
>On the other hand, with the current solution, axis_partitions and simple partitions should be interchangeable, which, it seems to me, is not what we want. We want the virtual partition obtained after re-partitioning to behave like a simple partition.

@anmyachev You bring up a good point, but I think we do want both column/row partitions to behave like simple partitions in the future. One of the optimizations we will need is pipelining, which will require us to be able to treat all modes of partitioning equally. Originally, the column/row (axis) partitions were meant to be transient, but in order to enable pipelining, we need to be able to store physical partitions in row/column partitioning. This is a step in that direction.

Reply:
Can you briefly describe what you mean by pipelining here? In what situations do we need that column/row partitions to behave like simple partitions?

Reply:
Sometimes it will not be efficient to split the data back into blocks after column/row axis partition. We have some description of what this looks like in this paper: https://people.eecs.berkeley.edu/~totemtang/paper/Modin.pdf

Reply:
Thanks! I will take a look.